pkg/kubecost/allocationprops.go

```diff
@@ -0,0 +1,308 @@
+
+	pAnnotations := p.Annotations
+	thatAnnotations := that.Annotations
+	if len(pAnnotations) != len(thatAnnotations) {
+		for k, pv := range pAnnotations {
+			tv, ok := thatAnnotations[k]
+			if !ok || tv != pv {
+				return false
+			}
+		}
+		return false
+	}
+
+	pServices := p.Services
+	thatServices := that.Services
+	if len(pServices) != len(thatServices) {
+		sort.Strings(pServices)
+		sort.Strings(thatServices)
+		for i, pv := range pServices {
+			tv := thatServices[i]
+			if tv != pv {
+				return false
+			}
+		}
+		return false
+	}
+
+	return true
+}
+
+func (p *AllocationProperties) String() string {
+	if p == nil {
+		return "<nil>"
+	}
+
+	strs := []string{}
+
+	if p.Cluster != "" {
+		strs = append(strs, "Cluster:"+p.Cluster)
+	}
+
+	if p.Node != "" {
+		strs = append(strs, "Node:"+p.Node)
+	}
+
+	if p.Container != "" {
+		strs = append(strs, "Container:"+p.Container)
+	}
+
+	if p.Controller != "" {
+		strs = append(strs, "Controller:"+p.Controller)
+	}
+
+	if p.ControllerKind != "" {
+		strs = append(strs, "ControllerKind:"+p.ControllerKind)
+	}
+
+	if p.Namespace != "" {
+		strs = append(strs, "Namespace:"+p.Namespace)
+	}
+
+	if p.Pod != "" {
+		strs = append(strs, "Pod:"+p.Pod)
+	}
+
+	if p.ProviderID != "" {
+		strs = append(strs, "ProviderID:"+p.ProviderID)
+	}
+
+	if len(p.Services) > 0 {
+		strs = append(strs, "Services:"+strings.Join(p.Services, ";"))
+	}
+
+	var labelStrs []string
+	for k, prop := range p.Labels {
+		labelStrs = append(labelStrs, fmt.Sprintf("%s:%s", k, prop))
+	}
+	strs = append(strs, fmt.Sprintf("Labels:{%s}", strings.Join(strs, ",")))
+
+	var AnnotationStrs []string
+	for k, prop := range p.Annotations {
+		AnnotationStrs = append(AnnotationStrs, fmt.Sprintf("%s:%s", k, prop))
+	}
+	strs = append(strs, fmt.Sprintf("Annotations:{%s}", strings.Join(strs, ",")))
+
+	return strings.Join(strs, ",")
+}
+
+// AggregationStrings converts a AllocationProperties object into a slice of strings
+// representing a request to aggregate by certain properties.
+// NOTE: today, the ordering of the properties *has to match the ordering
+// of the allocaiton function generateKey*
+func (p *AllocationProperties) AggregationStrings() []string {
+	if p == nil {
+		return []string{}
+	}
+
+	aggStrs := []string{}
+	if p.Cluster != "" {
+		aggStrs = append(aggStrs, AllocationClusterProp.String())
+	}
+	if p.Node != "" {
+		aggStrs = append(aggStrs, AllocationNodeProp.String())
+	}
+	if p.Container != "" {
+		aggStrs = append(aggStrs, AllocationContainerProp.String())
+	}
+	if p.Controller != "" {
+		aggStrs = append(aggStrs, AllocationControllerProp.String())
+	}
+	if p.ControllerKind != "" {
+		aggStrs = append(aggStrs, AllocationControllerKindProp.String())
+	}
+	if p.Namespace != "" {
+		aggStrs = append(aggStrs, AllocationNamespaceProp.String())
+	}
+	if p.Pod != "" {
+		aggStrs = append(aggStrs, AllocationPodProp.String())
+	}
+	if p.ProviderID != "" {
+		aggStrs = append(aggStrs, AllocationProviderIDProp.String())
+	}
```

Comment:
Perhaps I'm confusing things, but we're probably going to want to talk about ordering here.